### ISSUE TITLE
Some DeprecationWarnings don't show up in jupyter notebooks

### ISSUE THREAD

**Issue by huonw** [stellargraph/stellargraph#1117]
### Describe the bug

We have a few places that use `DeprecationWarning`s to flag things that are changing and help users migrate... however, these aren't being shown by default in scripts or Jupyter notebooks. The `DeprecationWarning` type only shows up for code in `__main__` by default (i.e. a script or REPL session or jupyter notebook ),  from https://docs.python.org/3/library/warnings.html#warning-categories

> `DeprecationWarning`: Base category for warnings about deprecated features when those warnings are intended for other Python developers (ignored by default, unless triggered by code in __main__).

We should be configuring our deprecation warnings to appear in those circumstances. Some options:
1. using the `stacklevel` parameter to ensure they point to the call of the deprecated function, not the call of `warnings.warn` 
   - this works for Jupyter for all Python versions, but not for the Python REPL or a script in Python 3.6, the default was changed in 3.7 https://www.python.org/dev/peps/pep-0565/, but we could add a `warnings.filterwarnings("once", category=DeprecationWarning, module="__main__")` call to our `__init__.py` file to override Python's defaults for Python 3.6 (probably not a good idea...)
   - (I think we should be doing this anyway, since problem is the caller, not the `warnings.warn` call itself)
2. switching to using `FutureWarning`, but, per that table, it's not quite right (https://github.com/ipython/ipython/issues/8478 is related, and mentions Pandas using `FutureWarning` so that it is shown by default)
3. similarly, doing what NumPy does, which is creating a custom warning class `VisibleDeprecationWarning` that isn't suppressed by default https://github.com/numpy/numpy/blob/965b41d418e6100c1afae0b6f818a7ef152bc25d/numpy/_globals.py#L44-L51 (we could potentially just use that, too)

I think we should do only 1, without changing the warnings filtering, because we're mostly focused on Jupyter notebooks, but I could easily be convinced of other options. For instance, the thinking with Python's defaults seems to be these deprecation (and "pending deprecation") warnings get caught by a test suite outside of `__main__` (and hence, not showing them by default is ok); for instance, pytest will capture and show all warnings by default. However, I imagine many users of StellarGraph won't have a test suite exercising their graph ML, since that seems to be relatively hard/rare in ML circle, and thus if they have separate files containing all the ML parts that are imported/run by a script or notebook they won't see these warnings. Therefore, 2 or 3 are tempting.

This was noticed in https://github.com/stellargraph/stellargraph/pull/1110#pullrequestreview-377320963.

### To Reproduce

1. Use a python shell or jupyter notebook to run some code that triggers a warning, like constructing from networks triggering https://github.com/stellargraph/stellargraph/blob/cdcde873f7677a25d8c18124ef88a599f31ca7d9/stellargraph/core/graph.py#L253-L256
   ```python
   import networkx as nx
   import stellargraph as sg
   sg.StellarGraph(nx.Graph())
   ```
2. Run the code after forcing all warnings to be shown `import warnings; warnings.simplefilter("always")`

### Observed behavior

1. No warnings
2. The deprecation warning is shown:
   ```
   .../stellargraph/core/graph.py:255: DeprecationWarning: Constructing a StellarGraph directly from a NetworkX graph has been replaced by the `StellarGraph.from_networkx` function
     DeprecationWarning,
   ```

### Expected behavior

The warning should be shown in both cases.

### Environment

Operating system: `Darwin-18.6.0-x86_64-i386-64bit`
Python version:
```
3.6.9 (default, Jul 10 2019, 12:25:55) 
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)]
```
Package versions:
<details>

```
absl-py==0.8.0
ansiwrap==0.8.4
appdirs==1.4.3
appnope==0.1.0
astor==0.8.0
atomicwrites==1.3.0
attrs==19.3.0
backcall==0.1.0
black==19.10b0
bleach==3.1.0
boto==2.49.0
boto3==1.9.230
botocore==1.12.230
cachetools==4.0.0
certifi==2019.9.11
chardet==3.0.4
Click==7.0
coverage==4.5.4
coveralls==1.8.2
cycler==0.10.0
decorator==4.4.0
defusedxml==0.6.0
docopt==0.6.2
docutils==0.15.2
entrypoints==0.3
gast==0.2.2
gensim==3.8.0
google-auth==1.10.0
google-auth-oauthlib==0.4.1
google-pasta==0.1.7
grpcio==1.23.0
h5py==2.10.0
idna==2.8
importlib-metadata==0.23
ipykernel==5.1.3
ipython==7.9.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
isodate==0.6.0
jedi==0.15.1
Jinja2==2.10.3
jmespath==0.9.4
joblib==0.13.2
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==5.3.4
jupyter-console==6.0.0
jupyter-core==4.6.1
Keras==2.2.5
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.0
kiwisolver==1.1.0
llvmlite==0.30.0
Mako==1.1.0
Markdown==3.1.1
MarkupSafe==1.1.1
matplotlib==3.1.1
mistune==0.8.4
more-itertools==7.2.0
mplleaflet==0.0.5
mypy==0.750
mypy-extensions==0.4.3
nbclient==0.1.0
nbconvert==5.6.1
nbformat==4.4.0
networkx==2.3
notebook==6.0.2
numba==0.46.0
numpy==1.17.2
oauthlib==3.1.0
opt-einsum==3.1.0
packaging==19.2
pandas==0.25.1
pandocfilters==1.4.2
papermill==1.2.1
parso==0.5.1
pathspec==0.6.0
pdoc3==0.7.2
pexpect==4.7.0
pickleshare==0.7.5
pluggy==0.13.1
prometheus-client==0.7.1
prompt-toolkit==2.0.10
protobuf==3.9.1
ptyprocess==0.6.0
py==1.8.0
py-cpuinfo==5.0.0
py4j==0.10.7
pyasn1==0.4.8
pyasn1-modules==0.2.7
pydot==1.4.1
Pygments==2.4.2
Pympler==0.8
pyparsing==2.4.2
pyrsistent==0.15.6
pyspark==2.4.4
pyspark-stubs==2.4.0.post6
pytest==5.3.1
pytest-benchmark==3.2.2
pytest-cov==2.8.1
pytest-repeat==0.8.0
python-dateutil==2.8.0
pytz==2019.2
PyYAML==5.1.2
pyzmq==18.1.1
qtconsole==4.6.0
rdflib==4.2.2
regex==2019.12.9
requests==2.22.0
requests-oauthlib==1.3.0
rsa==4.0
s3transfer==0.2.1
scikit-learn==0.21.3
scipy==1.4.1
seaborn==0.10.0
Send2Trash==1.5.0
six==1.12.0
smart-open==1.8.4
-e git+git@github.com:stellargraph/stellargraph.git@cdcde873f7677a25d8c18124ef88a599f31ca7d9#egg=stellargraph
tenacity==6.0.0
tensorboard==2.1.1
tensorflow==2.1.0
tensorflow-estimator==2.1.0
termcolor==1.1.0
terminado==0.8.3
testpath==0.4.4
textwrap3==0.9.2
toml==0.10.0
tornado==6.0.3
tqdm==4.42.1
traitlets==4.3.3
treon==0.1.3
typed-ast==1.4.0
typing-extensions==3.7.4.1
urllib3==1.25.3
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.15.6
widgetsnbextension==3.5.1
wrapt==1.11.2
zipp==0.6.0

```

</details>

### Additional context

N/A

**Comment by kjun9** [comment]
All options seem fine to me, but if I had to choose, my vote goes to option 3 - seems straightforward, customisable, and I think it'd be nice if the behaviour is consistent across all cases (notebook, REPL, script). I often just use a REPL for quick sanity checks.

**Comment by timpitman** [comment]
I agree that Option 3 sounds good, as Notebooks are out primary use-case and it's straightforward to implement.   Option 1 being better supported in 3.7 is fine; we'll eventually all move to 3.7+.

**Comment by timpitman** [comment]
Closing as already addressed by other PR's for python 3.7+ and Jupyter notebooks.
